dot/sync/service.go

```diff
@@ -0,0 +1,288 @@
+	s.wg.Wait()
+	return nil
+}
+
+func (s *SyncService) HandleBlockAnnounceHandshake(from peer.ID, msg *network.BlockAnnounceHandshake) error {
+	logger.Infof("receiving a block announce handshake: %s", from.String())
+	if err := s.workerPool.fromBlockAnnounceHandshake(from); err != nil {
+		return err
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	return s.currentStrategy.OnBlockAnnounceHandshake(from, msg)
+}
+
+func (s *SyncService) HandleBlockAnnounce(from peer.ID, msg *network.BlockAnnounceMessage) error {
+	gossip, repChange, err := s.currentStrategy.OnBlockAnnounce(from, msg)
+	if err != nil {
+		return fmt.Errorf("while handling block announce: %w", err)
+	}
+
+	if repChange != nil {
+		s.network.ReportPeer(repChange.rep, repChange.who)
+	}
+
+	if gossip {
+		s.network.GossipMessage(msg)
+	}
+
+	return nil
+}
+
+func (s *SyncService) OnConnectionClosed(who peer.ID) {
+	logger.Tracef("removing peer worker: %s", who.String())
+	s.workerPool.removeWorker(who)
+}
+
+func (s *SyncService) IsSynced() bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.currentStrategy.IsSynced()
+	return false
+}
+
+func (s *SyncService) HighestBlock() uint {
+	highestBlock, err := s.blockState.BestBlockNumber()
+	if err != nil {
+		logger.Warnf("failed to get the highest block: %s", err)
+		return 0
+	}
+	return highestBlock
+}
+
+func (s *SyncService) runSyncEngine() {
+	defer s.wg.Done()
+
+	logger.Infof("starting sync engine with strategy: %T", s.currentStrategy)
+
+	for {
+		select {
+		case <-s.stopCh:
+			return
+		default:
+		}
+
+		finalisedHeader, err := s.blockState.GetHighestFinalisedHeader()
+		if err != nil {
+			logger.Criticalf("getting highest finalized header: %w", err)
+			return
+		}
+
+		bestBlockHeader, err := s.blockState.BestBlockHeader()
+		if err != nil {
+			logger.Criticalf("getting best block header: %w", err)
+			return
+		}
+
+		logger.Infof(
+			"🚣 currently syncing, %d peers connected, finalized #%d (%s), best #%d (%s)",
+			len(s.network.AllConnectedPeersIDs()),
+			finalisedHeader.Number,
+			finalisedHeader.Hash().Short(),
+			bestBlockHeader.Number,
+			bestBlockHeader.Hash().Short(),
+		)
+
+		tasks, err := s.currentStrategy.NextActions()
+		if err != nil {
+			logger.Criticalf("current sync strategy next actions failed with: %s", err.Error())
+			return
+		}
+
+		if len(tasks) == 0 {
+			// sleep the amount of one slot and try
+			time.Sleep(s.slotDuration)
+			continue
+		}
+
+		results := s.workerPool.submitRequests(tasks)
+		done, repChanges, peersToIgnore, err := s.currentStrategy.IsFinished(results)
+		if err != nil {
+			logger.Criticalf("current sync strategy failed with: %s", err.Error())
+			return
+		}
+
+		for _, change := range repChanges {
+			s.network.ReportPeer(change.rep, change.who)
+		}
+
+		for _, block := range peersToIgnore {
+			s.workerPool.ignorePeerAsWorker(block)
+		}
+
+		s.currentStrategy.ShowMetrics()
+
+		if done {
+			if s.defaultStrategy == nil {
+				logger.Criticalf("nil default strategy")
```

Comment:
maybe this check can happen in construction of `SyncService`? This would be pretty annoying to get after you've finished the strategy.